core/src/main/scala/kafka/admin/AdminClient.scala

```diff
@@ -15,38 +15,54 @@ package kafka.admin
 import java.nio.ByteBuffer
 import java.util.{Collections, Properties}
 import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.{TimeUnit, Future}
 
+import kafka.admin.AdminClient.PurgeDataResult
 import org.apache.kafka.common.requests.ApiVersionsResponse.ApiVersion
 import kafka.common.KafkaException
 import kafka.coordinator.GroupOverview
 import kafka.utils.Logging
 import org.apache.kafka.clients._
-import org.apache.kafka.clients.consumer.internals.{ConsumerNetworkClient, ConsumerProtocol, RequestFuture}
+import org.apache.kafka.clients.consumer.internals.{RequestFutureAdapter, ConsumerNetworkClient, ConsumerProtocol, RequestFuture}
 import org.apache.kafka.common.config.ConfigDef.{Importance, Type}
 import org.apache.kafka.common.config.{AbstractConfig, ConfigDef}
 import org.apache.kafka.common.metrics.Metrics
 import org.apache.kafka.common.network.Selector
 import org.apache.kafka.common.protocol.{ApiKeys, Errors}
 import org.apache.kafka.common.requests._
 import org.apache.kafka.common.requests.OffsetFetchResponse
-import org.apache.kafka.common.utils.{Time, Utils}
+import org.apache.kafka.common.utils.{KafkaThread, Time, Utils}
 import org.apache.kafka.common.{Cluster, Node, TopicPartition}
 
 import scala.collection.JavaConverters._
 import scala.util.Try
 
 class AdminClient(val time: Time,
                   val requestTimeoutMs: Int,
+                  val retryBackoffMs: Long,
                   val client: ConsumerNetworkClient,
                   val bootstrapBrokers: List[Node]) extends Logging {
 
+  @volatile var running: Boolean = true
+
+  val networkThread = new KafkaThread("admin-client-network-thread", new Runnable {
+    override def run() {
+      while (running) {
+        client.poll(retryBackoffMs)
+        // Avoid busy loop so that user thread can get the lock on the ConsumerNetworkClient to enqueue requests.
+        if (client.pendingRequestCount() == 0)
+          Thread.sleep(retryBackoffMs)
```

Comment:
I have updated the AdminClient to remove the sleep operation. And the patch guarantees that if user thread attempts to send any request, it will get the lock on ConsumerNetworkClient and enqueue the requests once the currehnt `client.poll(..)` exits.

The user thread may still wait up to `maxRetryBackoffMs` in order to send the requests as of current implementation. This is because `ConsumerNetworkClient.poll(...)` does not guarantee that it will unblock and send the request if there is available request to be send by `ConsumerNetworkClient.send(...)`. The problem exists even if we have a queue outside of ConsumerNetworkClient and use only one thread to call `ConsumerNetworkClient.poll(...)` and `ConsumerNetworkClient.send(...)`. Having user thread wake up consumer may help reduce the chance of unnecessary block but doesn't prevent this from happening.

Ideally we want to be able to fix ConsumerNetworkClient so that user's request should get sent immediately if there is no other inflight requests. But it turns out to be a non-trivial fix. The problem is that `ConsumerNetworkClient.poll(...)` will hold the lock of `ConsumerNetworkClient` while it is blocked on `nioSelector.select(ms)`. If user calls `nioSelector.wakeup()`, it may be lost if the sender thread is right before the `nioSelector.select(ms)` while `nioSelector.wakeup()` is called. If user attempt to get the lock of `ConsumerNetworkClient` before calling `nioSelector.wakeup()`, then it will block as well while trying to get the lock. We need to find a way for the `poll(...)` to release lock while it is waiting on `nioSelector.select(ms)`. But I don't have an easy way to fix it.